vowpalwabbit/baseline_challenger_cb.cc

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "action_score.h"
+#include "debug_log.h"
+#include "reductions.h"
+#include "learner.h"
+#include <cfloat>
+
+#include "distributionally_robust.h"
+#include "io/logger.h"
+#include "vw.h"
+#include "example.h"
+
+#include <memory>
+#include <utility>
+#include <algorithm>
+
+using namespace VW;
+using namespace VW::config;
+using namespace VW::LEARNER;
+
+namespace logger = VW::io::logger;
+
+namespace VW
+{
+struct discounted_expectation {
+  discounted_expectation(double tau): tau(tau), sum(0), n(0) { }
+
+  void update(double w, double r) {
+    sum = tau * sum + w * r;
+    n = tau * n + w;
+  }
+
+  double current() const {
+    return n == 0 ? 0 : sum / n;
+  }
+
+  void save_load(io_buf& io, bool read, bool text, const char* name) {
+    std::stringstream msg;
+    if (!read) {
+      msg << name << "_expectation_sum " << std::fixed << sum << "\n";
+    }
+    bin_text_read_write_fixed_validated(
+        io, reinterpret_cast<char*>(&sum), sizeof(sum), "", read, msg, text);
+
+    if (!read) {
+      msg << name << "_expectation_n " << std::fixed << n << "\n";
+    }
+    bin_text_read_write_fixed_validated(
+        io, reinterpret_cast<char*>(&n), sizeof(sum), "", read, msg, text);
+  }
+private:
+  double tau, sum, n;
+};
+
+constexpr double DEFAULT_TAU = 0.999;
+constexpr double DEFAULT_ALPHA = 0.05;
+
+struct baseline_challenger_data {
+  distributionally_robust::ChiSquared baseline;
+  discounted_expectation policy_expectation;
+  float baseline_epsilon;
+  bool emit_metrics;
+
+  baseline_challenger_data(bool emit_metrics, float alpha, float tau) :
+      baseline(alpha, tau),
+      policy_expectation(tau),
+      emit_metrics(emit_metrics)
+  {
+  }
+
+  static int get_chosen_action(const ACTION_SCORE::action_scores& action_scores)
+  {
+    return action_scores[0].action;
+  }
+
+  template <bool is_learn>
+  inline void learn_or_predict(multi_learner& base, multi_ex& examples)
+  {
+    multiline_learn_or_predict<false>(base, examples, examples[0]->ft_offset);
+
+    //must not be const since we mutate it at the end
+    auto& action_scores = examples[0]->pred.a_s;
+    const uint32_t chosen_action = get_chosen_action(action_scores);
+
+    if (is_learn)
+    {
+      const auto lbl_example = std::find_if(examples.begin(), examples.end(), [](example* item) { return !item->l.cb.costs.empty(); });
+      const auto labelled_action = static_cast<uint32_t>(std::distance(examples.begin(), lbl_example));
+      if (lbl_example != examples.end()) {
```

Comment:
This check should be done before you cast to a uint32_t so it is still an iterator